Tests/StreamChatTests/Repositories/AuthenticationRepository_Tests.swift

```diff
@@ -0,0 +1,360 @@
+        XCTAssertNotNil(repository.currentUserId)
+
+        repository.clearTokenProvider()
+        XCTAssertNil(repository.tokenProvider)
+        XCTAssertNotNil(repository.currentToken)
+        XCTAssertNotNil(repository.currentUserId)
+    }
+
+    // MARK: Log out
+
+    func test_logOut_clearsUserData() {
+        let userInfo = UserInfo(id: "123")
+        repository.connectGuestUser(userInfo: userInfo, completion: { _ in })
+        repository.setToken(token: .unique())
+        XCTAssertNotNil(repository.tokenProvider)
+        XCTAssertNotNil(repository.currentToken)
+        XCTAssertNotNil(repository.currentUserId)
+
+        repository.logOutUser()
+        XCTAssertNil(repository.tokenProvider)
+        XCTAssertNil(repository.currentToken)
+        XCTAssertNil(repository.currentUserId)
+    }
+
+    // MARK: Refresh token
+
+    func test_refreshToken_returnsError_ifThereIsNoTokenProvider() {
+        let expectation = self.expectation(description: "Refresh token")
+        var receivedError: Error?
+        repository.refreshToken { error in
+            receivedError = error
+            expectation.fulfill()
+        }
+        waitForExpectations(timeout: 0.1)
+        XCTAssertTrue(receivedError is ClientError.MissingTokenProvider)
+    }
+
+    func test_refreshToken_success() throws {
+        try setTokenProvider(mockedResult: .success(.unique()))
+
+        let receivedError = try refreshTokenAndWaitForResponse(mockedError: nil)
+
+        XCTAssertCall("nextRetryDelay()", on: retryStrategy)
+        XCTAssertCall("resetConsecutiveFailures()", on: retryStrategy)
+        XCTAssertNil(receivedError)
+    }
+
+    func test_refreshToken_failure() throws {
+        let mockedError = ClientError("Mocked error")
+        try setTokenProvider(mockedResult: .failure(mockedError))
+
+        let receivedError = try refreshTokenAndWaitForResponse(mockedError: mockedError)
+
+        XCTAssertCall("nextRetryDelay()", on: retryStrategy)
+        XCTAssertNotCall("resetConsecutiveFailures()", on: retryStrategy)
+        XCTAssertEqual(receivedError, mockedError)
+    }
+
+    private func refreshTokenAndWaitForResponse(mockedError: Error?) throws -> Error? {
+        XCTAssertNotNil(repository.tokenProvider)
+        retryStrategy.mock_nextRetryDelay.returns(0.1)
+
+        let expectation = self.expectation(description: "Refresh token")
+        var receivedError: Error?
+        repository.refreshToken { error in
+            receivedError = error
+            expectation.fulfill()
+        }
+
+        AssertAsync.willBeTrue(clientUpdater.reloadUserIfNeeded_completion != nil)
+        let reloadUserCompletion = try XCTUnwrap(clientUpdater.reloadUserIfNeeded_completion)
+        let tokenProvider = try XCTUnwrap(clientUpdater.reloadUserIfNeeded_tokenProvider)
+        reloadUserCompletion(mockedError)
+        tokenProvider { _ in }
+        waitForExpectations(timeout: 10)
+        return receivedError
+    }
```

Comment:
I ideally we should follow top-to-bottom, where private and helper functions go to the bottom of the file.